Replace `Select` with `select()`, a type-safe async iterator

leandro-lucarella-frequenz opened the pull request:
This new function acts as an async iterator, and makes sure no dangling tasks are left behind after a select loop is done.
    
This class is designed to replace the current `Select` implementation with the following improvements:
    
* Proper type hinting by using the new helper type guard `selected_from()`.
* Fixes potential starvation issues.
* Simplifies the interface by providing values one-by-one.
* Simplifies the implementation, so it is easier to maintain.
* Guarantees there are no dangling tasks left behind.

Here is an usage example:

```python
import datetime

from frequenz.channels import ReceiverStoppedError
from frequenz.channels.util import select, selected_from, Timer

timer1 = Timer.periodic(datetime.timedelta(seconds=1))
timer2 = Timer.timeout(datetime.timedelta(seconds=0.5))

async for selected in select(timer1, timer2):
    if selected_from(selected, timer1):
        if selected.was_stopped():
            print("timer1 was stopped")
            continue
        print(f"timer1: now={datetime.datetime.now()} drift={selected.value}")
        timer2.stop()
    elif selected_from(selected, timer2):
        if selected.was_stopped():
            print("timer2 was stopped")
            continue
        if exception := selected.exception is not None:
            print("timer2 was had an error")
            continue
        print(f"timer2: now={datetime.datetime.now()} drift={selected.value}")
    else:
        # This is not necessary, as select() will check for exhaustiveness, but
        # it is good practice to have it in case you forgot to handle a new
        # receiver added to `select()` at a later point in time.
        assert False
```

llucax committed:
Add __str__ and __repr__ to Timer classes

This is to make it easier to debug and to quickly print a timer.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

shsms commented:
I get the feeling that in cases like this, where we are replacing the entire implementation of a component, we could get creative about how we make the commits, such that it is easy to review.

For example,  
1. we could remove the old Select in the first commit,  add the new one in the second commit
2. add SelectNew as a separate implementation in the commit,  that will just replace `Select` in the second commit,  so we can spend much less time reviewing the second commit.

Just a suggestion,  and it also won't always be so straightforward.

shsms commented on src/frequenz/channels/util/_select.py:
```diff
@@ -8,199 +8,223 @@
+                timeout=timeout.total_seconds() if timeout else None,
+            )
+            ready_set: set[_base_classes.Receiver[Any]] = set()
+            for task in done:
+                name = task.get_name()
+                recv = receivers_map[name]
+                # This will raise if there was an exception in the task, as this is
+                # unexpected (Receiver.ready() should not raise).
+                exc = task.exception()
+                if exc is not None:
+                    raise exc
+                ready_set.add(recv)
+
+            for ready in ready_set:
+                selected = Selected(ready)
+                yield selected
+                assert selected._handled, f"{selected} was not handled"
+
+            for task in done:
+                receiver_active = task.result()
+                if not receiver_active:
+                    continue
+                name = task.get_name()
+                recv = receivers_map[name]
+                pending.add(_asyncio.create_task(_recv_task(recv), name=name))
```
why not a single loop?  going over the same set of items 3 times is slower than one right?  and that we can also eliminate the need for a `ready_set`.

leandro-lucarella-frequenz replied:
It might be a leftover from a previous implementation.  At some point there was a reason for it, but maybe that reason doesn't exist any more.  I'll try to make all in one loop and if I can't I will comment on the why :+1:

llucax replied:
Removed the extra loops :rocket:

shsms commented on src/frequenz/channels/util/_select.py:
```diff
@@ -8,199 +8,223 @@
         )
+
+
+def is_selected(
```
I guess this will change to `selected_from`?

shsms reviewed (commented):
Thanks Luca,  just looked at the essential parts of the `_select.py` file and all that looks good to me, except for a perf and a naming comment.

leandro-lucarella-frequenz commented:
> I get the feeling that in cases like this, where we are replacing the entire implementation of a component, we could get creative about how we make the commits, such that it is easy to review.
> 
> For example,
> 
>     1. we could remove the old Select in the first commit,  add the new one in the second commit
> 
>     2. add SelectNew as a separate implementation in the commit,  that will just replace `Select` in the second commit,  so we can spend much less time reviewing the second commit.
> 
> 
> Just a suggestion, and it also won't always be so straightforward.

Yeah, sorry if it was unclear, but the commits are not clean at all, I just put the draft out there to have a look mainly on how it is used and the new select implementation.

llucax committed:
Make type hints use built-in types

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a new async iterable select() function

This function will replace the current `Select` implementation with the
following improvements:

* Proper type hinting by using the new helper type guard
  `selected_from()`.
* Fixes potential starvation issues.
* Simplifies the interface by providing values one-by-one.
* Simplifies the implementation, so it is easier to maintain.

There are some other improvements we would have liked to be able to make
but was difficult.

For example, typing for `select()` is tricky.  We had the idea of using
a declarative design, something like:

```python
class MySelector(Selector):
    receiver1: x.new_receiver()
    receiver2: y.new_receiver()

async for selected in MySelector:
    if selected.receiver is receiver1:
        # Do something with selected.value
    elif selected.receiver is receiver1:
        # Do something with selected.value
```

This is similar to `Enum`, but `Enum` has special support in `mypy` that
we can't have.

With the current implementation, the typing could be slightly improved
by using `TypeVarTuple`, but we are not because "transformations" are
not supported yet, see: https://github.com/python/typing/issues/1216

Also support for `TypeVarTuple` in general is still experimental (and
very incomplete in `mypy`).

With this we would also probably be able to properly type `select` and
*maybe* even be able to leverage the exhaustiveness checking of `mypy`
to make sure the selected value is narrowed down to the correct type to
make sure all receivers are handled, with the help of `assert_never` as
described in:
https://docs.python.org/3.11/library/typing.html#typing.assert_never

We also explored the possibility of using `match` to perform
exhaustiveness checking, but we couldn't find a way to make it work
with `match`, and `match` is not yet checked for exhaustiveness by
`mypy` anyway, see: https://github.com/python/mypy/issues/13597

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
I updated the PR:

* Addressed @shsms comments (including better splitting of commits)
* The implementation itself now can be considered final, so it would be good to have a final review before I get into adding tests and replacing the current `Select` use in all the code.

TODO:
* Add tests
* Replace usage of the old `Select` with the new `select`
* Remove the old `Select`

llucax committed:
Add a Event utility receiver

This receiver can be manually made ready.  It is mainly useful for
testing but can also become handy in scenarios where a simple, on-off
signal needs to be sent to a select loop for example.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove unused __future__.annotations

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove type hints from docstrings

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update existing code to use Selector instead of Select

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add test for the Selector class

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update issue template to use "selector" instead of "select"

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Rename the util._select module to util._selector

This is to match the main contents of the module, the `Selector` class.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Revert "Add a new async iterable select() function"

Sadly the generator approach doesn't work because of limitations in how
Python handles the `yield` statement and breaking out from a `async for`
for loop (actually this applies to sync generators too).

When the loop is broken, the control is never passed back to the async
genreator, so the `finally` block is never executed (at least not until
the end of the program, where dangling generators are cleared).

Because of this, we will need to use a class instead, and to make it
easy to guarantee avoiding leaking resources (tasks), we make it an
async context managaer and an async iterator.

This reverts commit 2764d0171d926bc2e41c29eca45a719fcc797f3f.

Example:

```python
import asyncio
from typing import AsyncIterator

was_gen_finalized: bool = False

async def gen() -> AsyncIterator[int]:
    try:
        print("gen")
        yield 1
    finally:
        global was_gen_finalized
        was_gen_finalized = True
        print("gen finally")

async def main() -> None:
    global was_gen_finalized
    print("1. without break")
    async for i in gen():
        print(i)
    print(f"  end of loop: {was_gen_finalized=}")
    was_gen_finalized = False

    print("------------------")
    print("2. with break")
    async for i in gen():
        print(i)
        break

    print(f"  end of loop: {was_gen_finalized=}")
    was_gen_finalized = False

    print("------------------")
    print("2. with exception")
    try:
        async for i in gen():
            print(i)
            raise Exception("Interrupted by exception")
    except Exception:
        pass

    print(f"  end of loop: {was_gen_finalized=}")
    was_gen_finalized = False

    print()
    print("END of main")

asyncio.run(main())
print(f"END of asyncio.run(): {was_gen_finalized=}")
```

This program prints:
```
1. without break
gen
1
gen finally
  end of loop: was_gen_finalized=True
------------------
2. with break
gen
1
  end of loop: was_gen_finalized=False
------------------
2. with exception
gen
1
  end of loop: was_gen_finalized=False

END of main
gen finally
gen finally
END of asyncio.run(): was_gen_finalized=True
```

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a new Selector class

This new class acts as an async context manager and an async iterator,
and makes sure no dangling tasks are left behind after a select loop is
done.

This class (as the failed `select()` function) is also designed to
replace the current `Select` implementation with the following
improvements:

* Proper type hinting by using the new helper type guard
  `selected_from()`.
* Fixes potential starvation issues.
* Simplifies the interface by providing values one-by-one.
* Guarantees there are no dangling tasks left behind when used as an
  async context manager.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove the old Select implementation

This class was replaced by the new Selector class.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
OK, here we go!  I wish this would not have taken so much time to get right (hopefully this time is right :crossed_fingers: ).

I decided to leave the failed `select()` function implementation for *historical reasons*. Also because at some point I doubted if it was a viable approach or not and even when now I'm pretty sure it is not, I wonder if Python will eventually change this. It is very weird, for one side it makes a lot of sense that a generator never regains control after a `async for` loop is broken, but for the other side it looks super broken that a `finally` statement can be skipped, so who knows...

llucax commented:
Oh, I need to add release notes :spiral_notepad:

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Added release notes.

shsms commented on src/frequenz/channels/util/_select.py:
```diff
@@ -11,13 +11,598 @@
+# https://github.com/python/mypy/issues/13597
+
+
+class Selector:
```
I think this should be renamed to `Select` because it is not used like a regular Python class which we create and call methods on.  It is used like a select function,  and is implemented as a class because of Python language limitations.  And it would then also be the same idiom from other languages.

llucax replied:
I named it `Selector` because for me it is still mainly an async generator, not really a function, there is even no function call syntax involved to use it (except for creating the object). Even with the previous implementation it was weird to me to call it `Select`, because if was even more an object than this implementation, as you did have to call methods (like `ready()` on it).

Naming discussions are complicated and at the end I think the goal is to use whatever feels more obvious to users, so I will create a poll in our internal chat to see what people think is more readable/obvious, hope that is OK with you.

llucax replied:
`select()` is back! The name discussion is moot now :tada:

shsms reviewed (dismissed):
Just a comment about the name,  which is of course subjective,  but looks good otherwise.

llucax committed:
Add a new async generator implementation for select()

It turns out the *broken* async generators are not completely leaked, it
just needs more "time" to be cleaned up.

In the example in 9015e7da0f9faeb2bae470d62c040f9ea1839ceb if a few
`await asyncio.sleep(0)` are added, then the generator is properly
finalized. Even when the destruction is much less deterministic than
using a context manager, it should be safe enough to assume that in a
long running program the generator will be eventually (and very soon)
cleaned up, and the user interface is much simpler and intuitive, so
we decided to bring it back.

This in a way reverts 9015e7da0f9faeb2bae470d62c040f9ea1839ceb but it
also simplifies the code and fixes a few bugs.

The tests of the selector are ported to test select() and we make sure
all tasks are properly cleaned up after a few yields to the event loop.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Sort exported symbols

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Port all existing uses of Selector to select

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove the Selector class and tests

This was replaced by the `select()` function.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Renames files from `selector` to `select`

Since the `Selector` class is gone, it makes sense to name files as
`select` again.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Updated to bring `select()` back! What a ride! :nauseated_face:

llucax commented:
Enabled auto-merge :scream:
